doc/styleguide.md

```diff
@@ -122,3 +122,43 @@ it.
 ## Function and variable names
 
 See [API and naming](api-and-naming.md).
+
+## Quoting of words containing shell expansions such as `$var`, `$(cmd)`, and `$((expr))`
+
+To avoid unexpected word splitting and pathname expansions, an argument of
+commands needs to be properly quoted when it contains shell expansions.
+
+When one intentionally wants word splitting and pathname expansions, one should
+consider using the utility functions provided by bash-completion.  To safely
+split a string without being affected by word splitting and pathname
+expansions, use the shell function `_comp_split`.  To safely obtain filenames
+by pathname expansions without being affected by `failglob`, etc., use the
+shell function `_comp_expand_glob`.  Note that `_comp_expand_glob` should be
+always used for the pathname patterns even if the pattern does not contain
+shell expansions.
+
+In the following contexts, the quoting to suppress word splitting and pathname
+expansions are not needed.
+
+- The right-hand sides of variable assignments ... `v=$var`
+- The arguments of conditional commands ... `[[ $var ]]`
+- The argument specified to `case` statement ... `case $var in foo) ;; esac`
+
+However, there are exceptions where the quoting is still needed.
+
+- When the word contains shell special characters ``/[ \t\n;|&()<>\\$`'"#!~{]/``,
```

Comment:
What is "the word" here, could we rephrase/clarify this some, perhaps through an example?

Reply:
Hmm. The "word" here means the range of characters in the shell code that the user wants to treat as a single shell WORD. WORD is defined in [POSIX XCU 2.3 Token Recognition](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_03) and in [XCU 2.10.2 Shell Grammar Rules](https://pubs.opengroup.org/onlinepubs/9699919799/). So, if the text range contains unquoted shell special characters, that is actually no longer a shell word. In that sense, the word cannot contain these characters without quoting.

Reply:
Sorry, I was probably unclear, let me try to elaborate. The bullet point examples above contain uses of `$var` (which are not "word"s AFAIU), but this bullet talks about exceptions, and I read them being exceptions to the above points. But because the above points do not contain any "word" uses -> confusion ensues exactly to which are these exceptions for.

So reading this I'm left wondering if quoting is necessary e.g. (using the slash as an example) for literal `foo/bar` so we'd need to do `v="foo/bar"`, or uses of variables whose values might contain the slash (in which case we might not know, and it would arguably to be better to just always quote).

Reply:
Thank you for your clarification! I still planned to update the corresponding part, but I was thinking about how I should fix it. Now I understand the problems.

In my mind, I was talking about "the words that *contain* shell expansions" in the entire section as in the section title, and `$var` was the simplest case of such a word. However, as you pointed out, the examples seem to just talk about the shell expansions so are confusing. Then, I guess adding examples would be fine to avoid the confusion.

> So reading this I'm left wondering if quoting is necessary e.g. (using the slash as an example) for literal `foo/bar` so we'd need to do `v="foo/bar"`,

Sorry for confusing you, but the slash is not a part of the set of special characters. The enclosing `[` and `]` are also not a part of the special characters. I have a habit of indicating a regular expression by `/.../`. (Actually, I initially wrote `` \t\n;|&()<>\\$`'"#!~{`` without `/[` and `]/`, but the pre-commit hook (probably PyMarkdown) has complained that a whitespace character cannot be included at the beginning of the code span.) I'll try to fix it.

Reply:
ca0fde09

- I changed the section title to `Quoting of words` to extend the target to the general words.
- I made it explicit that we avoid unnecessary quoting.
- I explained the exception of shell special characters in more detail.
- I added the case of an empty word as an exception.
- I also added our preferences in the type of quoting as suggested in [#932 (comment)](https://github.com/scop/bash-completion/issues/932#issuecomment-1556879298).

Reply:
Looks great to me :+1: